drivers/platform/aducm3029/spi.c

```diff
@@ -0,0 +1,232 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include "spi_extra.h"
+#include "spi.h"
+#include "error.h"
+#include <stdlib.h>
+
+/******************************************************************************/
+/*****************************  Variables   **********************************/
+/******************************************************************************/
+
+static uint8_t	active_spi = 0; //Stores the active SPIs
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Allocates the memory needed for the descriptor
+ * @return A address to the allocated memory, NULL if the allocation fails.
+ */
+struct spi_desc *alloc_data_mem()
+{
+	struct spi_desc		*desc;
+	struct adicup_spi_desc	*extra;
+
+	desc = malloc(sizeof(struct spi_desc));
+	if (NULL == desc)
+		return NULL;
+
+	extra = malloc(sizeof(struct adicup_spi_desc));
+	if (NULL == extra) {
+		free(desc);
+		return NULL;
+	}
+	desc->extra = extra;
+
+	extra->buffer = malloc(ADI_SPI_MEMORY_SIZE);
+	if (NULL == extra->buffer) {
+		free(extra);
+		free(desc);
+		return NULL;
+	}
+
+	return desc;
+}
+
+/**
+ * @brief Deallocates the memory needed for the descriptor
+ * @param desc - Descriptor to be deallocated
+ */
+void free_desc_mem(struct spi_desc *desc)
+{
+	free(((struct adicup_spi_desc*)(desc->extra))->buffer);
+	free(desc->extra);
+	free(desc);
+}
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param init_param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_init(struct spi_desc **desc,
+		 const struct spi_init_param *param)
+{
+	ADI_SPI_RESULT			spi_ret;
+	struct adicup_spi_desc		*adicup_desc;
+	struct adicup_spi_init_param	*adicup_init_param;
+
+
+	if (desc == NULL || param == NULL)
+		goto fail_no_mem;
+
+	/* Verify if the SPI is already activated */
+	adicup_init_param = param->extra;
+	if (active_spi & (1 << adicup_init_param->spi_channel))
+		goto fail_mem;
```

Comment:
you don't have memory allocated yet.

Reply:
_param_ is allocated by the user and _adicup_init_param_ is used only for readability to avoid long casting